### ISSUE TITLE
Bug in Masked Patch Prediction ? 

### ISSUE THREAD

**Issue by Siyuan89** [lucidrains/vit-pytorch#97]
Hi, 

It seems that the label generation for Masked Patch Prediction are wrong. The loss converges to 0 after a couple of iterations as the generated labels remain constant and the same in all and every iterations. Here's an example of generated labels:

```
tensor([292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292,
        292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292,
        292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292,
        292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292,
        292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292,
        292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292,
        292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292,
        292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292,
        292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292,
        292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292,
        292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292,
        292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292,
        292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292,
        292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292, 292,
        292, 292, 292, 292])
```

I appreciate if you can look into this.

**Comment by lucidrains** [comment]
@zankner pulling in Zack :)

**Comment by zankner** [comment]
@lucidrains @Siyuan89 Thanks for reporting this. I'll take a look at it today.

**Comment by Siyuan89** [comment]
Thank you @zankner . Just for reference, reproducing can be done using this snippet: 

```
import torch
from vit_pytorch_source import ViT
from vit_pytorch_source.mpp import MPP

model = ViT(
    image_size=256,
    patch_size=32,
    num_classes=1000,
    dim=1024,
    depth=6,
    heads=8,
    mlp_dim=2048,
    dropout=0.1,
    emb_dropout=0.1
)

mpp_trainer = MPP(
    transformer=model,
    patch_size=32,
    dim=1024,
    mask_prob=0.15,          # probability of using token in masked prediction task
    random_patch_prob=0.30,  # probability of randomly replacing a token being used for mpp
    replace_prob=0.50,       # probability of replacing a token being used for mpp with the mask token
)

opt = torch.optim.Adam(mpp_trainer.parameters(), lr=3e-4)

def sample_unlabelled_images():
    return torch.randn(20, 3, 256, 256)


for _ in range(100):
    images = sample_unlabelled_images()
    loss = mpp_trainer(images)
    print(loss)
    opt.zero_grad()
    loss.backward()
    opt.step()

```

**Comment by zankner** [comment]
@Siyuan89 @lucidrains Hey, so I don't think its a bug. I think its just a result of the demo code. The `sample_unlabelled_images ` function just returns samples from the normal distribution. This makes all the values relatively small, and the target image has its pixels discretized which starts with a binning process, so all the values fall into the first bin which results in the seemingly constant target vals. If you just change the function to generate images from a different distribution you will notice different results. If I can clarify anything just let me know.

**Comment by Siyuan89** [comment]
Thanks for looking into it. I tried it with ImageNet data.. still getting the same masks at every iterations and converging almost instantly to zero. Have you been able to test it on actual data yourself as well ?

**Comment by lucidrains** [comment]
@zankner do you have a graph from one of your previous test runs on cifar?

**Comment by zankner** [comment]
@lucidrains I don't have any graphs, don't have a computer that can train ViT. But I did run it for a few batches of CIFAR and looked at the targets and they weren't constant. I think it's that I didn't normalize my images, just scaled them to 0-1. With normalization, your values will be centered around 0 and will mostly fall into the first bucket which would make the discrete target appear constant. I'm working on fixing MPP so that someone can just pass in the mean and std they normalized the inputs with, and the bins will be calculated accordingly.

**Comment by zankner** [comment]
#101 possible fix

**Comment by lucidrains** [comment]
@zankner you can actually train in the browser, using https://colab.research.google.com/ !

**Comment by zankner** [comment]
> @zankner you can actually train in the browser, using https://colab.research.google.com/ !

Oh, nice! Didn't know collab could handle it. I did get the chance to partially test it locally and on my end, the constant token issue went away after adding the change

**Comment by lucidrains** [comment]
sounds good, let's have @Siyuan89 try it out!

**Comment by Siyuan89** [comment]
@lucidrains @zankner Thank you for your prompt attention and possible fix solution. I will try it out on CIFAR and let you know !
